console/collections/src/merkle_tree/mod.rs

```diff
@@ -91,9 +97,28 @@ impl<E: Environment, LH: LeafHash<Hash = PH::Hash>, PH: PathHash<Hash = Field<E>
             // Compute the end index of the current level.
             let end = left_child(start);
             // Construct the children for each node in the current level.
-            let tuples = (start..end).map(|i| (tree[left_child(i)], tree[right_child(i)])).collect::<Vec<_>>();
+            let tuples = (start..end)
+                .filter_map(|i| {
+                    // Procure the children of the node.
+                    let tuple = (tree.get(left_child(i)).copied(), tree.get(right_child(i)).copied());
+                    // If both children are missing, return `None`.
+                    if tuple.0.is_none() && tuple.1.is_none() {
+                        None
+                    } else {
+                        Some((tuple.0.unwrap(), tuple.1.unwrap()))
```

Comment:
Given that we return a `Result`, I think we might as well `ensure` that the left and right tuple are both `Some` in order to prevent a panic.

Reply:
Actually, it's impossible for only one of them to be `None`, because we pad single-child nodes with a zero hash; however, this means that we might as well only check the second tuple, which will be faster :+1:

Reply:
Also, this comment made me realize that we could actually use `take_while` instead of `filter_map` here for shorter iteration :eyes:.